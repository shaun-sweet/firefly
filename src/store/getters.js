export default {
  user (state) {
    return {
      displayName: state.displayName,
      email: state.email,
      id: state.id
    }
  },

  homesList (state) {
    return Object.keys(state.homes).map(homeId => {
      return {
        label: state.homes[homeId].nickname,
        value: homeId
      }
    })
  },

  selectedHome (state) {
    return state.selectedHome
  },

  isLoadingInitialState (state) {
    return state.appState.isLoadingInitialState
  },

  defaultHome (state) {
    return Object.keys(state.homes).filter(home => state.homes[home].role === 'owner')[0]
  },

  devicesViewList (state, getters) {
    if (!getters.isLoadingInitialState) {
      const devicesViewList = state.homes[state.selectedHome].devicesViewList
      const devicesArray = Object.keys(devicesViewList).map((deviceId) => {
        const device = devicesViewList[deviceId]
        // this is what auto complete fuzzy finder search feature needs to display the properties
        device.stamp = device.primaryStateStatus
        device.label = device.alias
        return device
      })
      return devicesArray.filter((device) => device.export_ui)
    }
    return []
  }

  // notifications (state, getters) {
  //   const selectedHome = state.homes[getters.selectedHome]
  //   return Object.keys(selectedHome.notifications).map(id => selectedHome.notifications[id])
  // },

  // events (state, getters) {
  //   const selectedHome = state.homes[getters.selectedHome]
  //   return Object.keys(selectedHome.events).map(id => selectedHome.events[id])
  // }
}
