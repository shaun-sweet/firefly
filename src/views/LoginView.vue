<template>
  <section @keyup.enter.prevent="login()">
    <div class="layout-padding">
      <form>
        <q-field>
          <q-input icon="ion-email" type="email" stack-label="Email Address" placeholder="Just click Login for dev purposes" v-model="email" />
        </q-field>
        <q-field>
          <q-input v-model="password" stack-label="Password" type="password">
          </q-input>
        </q-field>
        <q-field>
          <q-btn color="primary" icon="ion-person" @click.prevent="login()">Login</q-btn>
          <q-btn color="primary" icon="ion-person" @click="handleLogout()">Logout</q-btn>
          <q-btn outline color="primary" icon="ion-person-add"> Signup</q-btn>
          <q-btn outline color="primary" icon="ion-ios-loop"> Send Password Reset</q-btn>
          <q-btn outline color="primary" icon="ion-checkmark-round">Send Email Verification</q-btn>
        </q-field>
        <p class="status" v-if="isLoggedIn">Firefly Sign-in Status: {{userEmail}} online</p>
      </form>
    </div>
  </section>
</template>

<script>
import { attemptLogin } from 'src/firebase'
import { Toast } from 'quasar'
import { ERROR_LOGIN } from 'src/store/mutation-types'

export default {
  data () {
    return {
      userEmail: '',
      errorMessage: '',
      isLoggedIn: false,
      email: '',
      password: ''
    }
  },
  methods: {
    handleLogout () {
      this.$store.dispatch('logOut')
      this.isLoggedIn = false
    },
    login () {
      // TODO: for dev only, remove for production
      let email = 'sweetone06@gmail.com'
      let password = 'destroy'
      if (this.email && this.password) {
        email = this.email
        password = this.password
      }
      attemptLogin(email, password)
        .then((response) => {
          this.isLoggedIn = true
          this.userEmail = response.email
          Toast.create.positive({
            html: `Successfully Logged in!  <br />
              Welcome back ${this.userEmail}`,
            icon: 'ion-ios-checkmark'
          })
        }, (err) => {
          const errorMsg = err.message
          this.isLoggedIn = false
          this.$store.commit(ERROR_LOGIN, { errorMsg })
          this.errorMessage = this.$store.state.errors.login

          Toast.create.negative({
            html: this.errorMessage,
            icon: 'ion-android-hand'
          })
        })
    }
  }
}
</script>

<style lang="stylus" scoped>
.status
  margin-top 1em
</style>
