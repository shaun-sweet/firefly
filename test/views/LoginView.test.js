import { shallow, createLocalVue } from 'vue-test-utils'
import Vuex from 'vuex'
import LoginView from 'src/views/LoginView.vue'
import getters from 'src/store/getters'
import {initialState, testState} from '../fixtures/state'

const localVue = createLocalVue()

localVue.use(Vuex)

describe('LoginView', () => {
  let wrapper;
  let store;
  let actions;
  let mutations;
  let state;
  let spy;
  beforeEach(() => {
    spy = jest.fn()
    state = {...testState}
    actions = {
      logOut: jest.fn(),
      login: jest.fn()
    }
    mutations = {
      ERROR_LOGIN: jest.fn()
    }
    store = new Vuex.Store({
      state,
      actions,
      mutations
    })
    wrapper = shallow(LoginView,{ store, localVue });
  })
  it('should render the LoginView correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should call handleLogout and dispatch action', () => {
    store.hotUpdate({
      actions: { logOut: spy }
    })
    wrapper.findAll('q-btn').at(1).trigger('click');
    expect(spy).toHaveBeenCalled()
  });

  it('should call login and dispatch action', () => {
    const errMsg = "test error"
    const email = 'test@test.com'
    const password = 'password'
    // TODO: Potentially to mock out attemptLogin on firebase
    // spy = jest.fn((email, password) => {
    //   return new Promise((resolve, reject) => {
    //     if( email && password === 'password'){
    //       return resolve(email)
    //     } else {
    //       reject(errMsg)
    //     }
    //   })
    // })
    spy = jest.fn()
    wrapper.vm.login = spy
    wrapper.update()
    wrapper.setData(email, password)
    wrapper.findAll('q-btn').at(0).trigger('click');
    expect(spy).toHaveBeenCalled();
  });

})
