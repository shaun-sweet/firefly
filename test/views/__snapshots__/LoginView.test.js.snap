// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginView should render the LoginView correctly 1`] = `
<section>
  <div
    class="layout-padding"
  >
    <form>
      <q-field>
        <q-input
          icon="ion-email"
          stack-label="Email Address"
          type="email"
        />
      </q-field>
       
      <q-field>
        <q-input
          stack-label="Password"
          type="password"
        />
      </q-field>
       
      <q-field>
        <q-btn
          color="primary"
          icon="ion-person"
        >
          Login
        </q-btn>
         
        <q-btn
          color="primary"
          icon="ion-person"
        >
          Logout
        </q-btn>
         
        <q-btn
          color="primary"
          icon="ion-person-add"
          outline=""
        >
           Signup
        </q-btn>
         
        <q-btn
          color="primary"
          icon="ion-ios-loop"
          outline=""
        >
           Send Password Reset
        </q-btn>
         
        <q-btn
          color="primary"
          icon="ion-checkmark-round"
          outline=""
        >
          Send Email Verification
        </q-btn>
      </q-field>
       
      <!---->
    </form>
  </div>
</section>
`;
